map/traffic_manager.cpp

```diff
@@ -158,56 +162,65 @@ void TrafficManager::Invalidate()
   if (!IsEnabled())
     return;
 
-  m_lastMwmsByRect.clear();
+  m_lastDrapeMwmsByRect.clear();
+  m_lastRoutingMwmsByRect.clear();
 
   if (m_currentModelView.second)
     UpdateViewport(m_currentModelView.first);
   if (m_currentPosition.second)
     UpdateMyPosition(m_currentPosition.first);
 }
 
-void TrafficManager::UpdateViewport(ScreenBase const & screen)
+void TrafficManager::UpdateActiveMwms(m2::RectD const & rect,
+                                      vector<MwmSet::MwmId> & lastMwmsByRect,
+                                      set<MwmSet::MwmId> & activeMwms)
 {
-  m_currentModelView = {screen, true /* initialized */};
-
-  if (!IsEnabled() || IsInvalidState())
+  auto mwms = m_getMwmsByRectFn(rect);
+  if (lastMwmsByRect == mwms)
     return;
-
-  if (df::GetZoomLevel(screen.GetScale()) < df::kRoadClass0ZoomLevel)
-    return;
-
-  // Request traffic.
-  auto mwms = m_getMwmsByRectFn(screen.ClipRect());
-  if (m_lastMwmsByRect == mwms)
-    return;
-  m_lastMwmsByRect = mwms;
+  lastMwmsByRect = mwms;
 
   {
     lock_guard<mutex> lock(m_mutex);
-
-    m_activeMwms.clear();
+    activeMwms.clear();
     for (auto const & mwm : mwms)
     {
       if (mwm.IsAlive())
-        m_activeMwms.insert(mwm);
+        activeMwms.insert(mwm);
     }
-
     RequestTrafficData();
   }
 }
 
 void TrafficManager::UpdateMyPosition(MyPosition const & myPosition)
 {
+  // Side of square around |myPosition|. Every mwm which is covered by the square
+  // will get traffic info.
+  double const kSquareSideM = 5000;
   m_currentPosition = {myPosition, true /* initialized */};
 
   if (!IsEnabled() || IsInvalidState())
     return;
 
-  // 1. Determine mwm's nearby "my position".
+  m2::RectD const rect = MercatorBounds::RectByCenterXYAndSizeInMeters(myPosition.m_position, kSquareSideM);
+  // Request traffic.
+  UpdateActiveMwms(rect, m_lastRoutingMwmsByRect, m_activeRoutingMwms);
```

Comment:
Why do we need to keep separate sets of mwms? I guess there should be a single set of mwms we are interested in, and we require traffic info for the set.